Add :downcase_request_headers option to HTTP1.connect

DunyaKokoschka opened the pull request:
i hear you want not lower header. i give you not lower header. but also i make default headers not lower if you pass
 in not lower option. alexi say this change is slow. i tell alexi if you care about slow why using elixir.

https://github.com/elixir-mint/mint/issues/256

whatyouhide commented:
Do you know if this is a behavior that we should have for HTTP/2 as well?

ericmj commented:
> Do you know if this is a behavior that we should have for HTTP/2 as well?

I would say no until proven otherwise. AFAIK all HTTP/2 implementers properly handle header casing.

ericmj commented on lib/mint/http1/headers.ex:
```diff
@@ -0,0 +1,61 @@
+  end
+
+  # name is required to be downcase ascii
+  def contains_header(headers, name) do
```
```suggestion
  def has_header?(headers, name) do
```

ericmj commented on lib/mint/http1/headers.ex:
```diff
@@ -0,0 +1,61 @@
+
+  # name is required to be downcase ascii
+  # downcase_ascii(new_name) == name
+  def replace_header(headers, name, new_name, value) do
```
Let's call the parameter `canonical_name` and use the same parameter order as `put_new_header`.

ericmj commented on lib/mint/http1/headers.ex:
```diff
@@ -0,0 +1,61 @@
+  # format of headers is [{header_name, canonical_header_name, value}]
+
+  def from_raw_headers(headers) do
+    for {name, value} <- headers, do: {name, Util.downcase_ascii(name), value}
```
We can skip calling `Util.downcase_ascii/1` by checking `downcase_headers` like in `to_raw_headers/2`.

DunyaKokoschka replied:
We need to store the canonical name in both cases whether we are preserving header casing or not because some of the logic requires the canonical name in order to work correctly. For example `replace_header` or `put_new_header` or `has_header?` requires knowledge of the canonical name.

ericmj commented on test/mint/http1/conn_test.exs:
```diff
@@ -650,6 +650,120 @@ defmodule Mint.HTTP1Test do
+          conn,
+          "GET",
+          "/",
+          [{"User-Agent", "myapp/1.0"}, {"Host", "localhost"}, {"Transfer-Encoding", "gzip"}],
```
```suggestion
          [{"User-Agent", "myapp/1.0"}, {"Host", "localhost"}, {"TRANSFER-ENCODING", "gzip"}],
```

Let's test this to check that we preserve the user's casing.

ericmj commented on lib/mint/http1.ex:
```diff
@@ -192,7 +194,8 @@ defmodule Mint.HTTP1 do
         state: :open,
-        log: log?
+        log: log?,
+        downcase_request_headers: Keyword.get(opts, :downcase_request_headers, true)
```
The option name should be `:case_sensitive_headers` and please also add docs for it.

whatyouhide replied:
I’m leaning towards a "verb" name too. `:case_sensitive_headers`, IMO, doesn't really convey the fact that this option controls whether we *downcase* headers. What do you think of `:preserve_header_casing`?

@ericmj also, I would really truly love to have a really awkward name for this, like `:preserve_non_compliant_header_casing` or something like this. How do you feel about that?

whatyouhide commented on lib/mint/http1/headers.ex:
```diff
@@ -0,0 +1,61 @@
+defmodule Mint.HTTP1.Headers do
+  alias Mint.Core.Util
+
+  # format of headers is [{header_name, canonical_header_name, value}]
```
Instead of having this as a comment, let's add it as a type so we can also use it for specs:

```suggestion
  @type header() :: {original_name :: String.t(), canonical_name :: String.t(), value :: String.t()}
  @type raw_header() :: {original_name :: String.t(), value :: String.t()}
```

whatyouhide commented on lib/mint/http1/headers.ex:
```diff
@@ -0,0 +1,61 @@
+  # format of headers is [{header_name, canonical_header_name, value}]
+
+  def from_raw_headers(headers) do
+    for {name, value} <- headers, do: {name, Util.downcase_ascii(name), value}
```
Let's not use a `for` comprehension here, because it will silently skip elements of `headers` that don't look like `{name, value}`.

```suggestion
    Enum.map(headers, fn {name, value} ->
      {name, Util.downcase_ascii(name), value}
    end)
```

(plus what @ericmj said)

whatyouhide commented on lib/mint/http1/headers.ex:
```diff
@@ -0,0 +1,61 @@
+
+  # format of headers is [{header_name, canonical_header_name, value}]
+
+  def from_raw_headers(headers) do
```
```suggestion
  @spec from_raw_headers([raw_header()]) :: [header()]
  def from_raw_headers(headers) when is_list(headers) do
```

whatyouhide commented on lib/mint/http1/headers.ex:
```diff
@@ -0,0 +1,61 @@
+
+  def to_raw_headers(headers, downcase_headers) do
+    if downcase_headers do
+      for {_name, canonical_name, value} <- headers do
+        {canonical_name, value}
+      end
+    else
+      for {name, _canonical_name, value} <- headers, do: {name, value}
```
Same with the `for` as above here. Use `Enum.map/2`.

whatyouhide commented on lib/mint/http1/headers.ex:
```diff
@@ -0,0 +1,61 @@
+    for {name, value} <- headers, do: {name, Util.downcase_ascii(name), value}
+  end
+
+  def to_raw_headers(headers, downcase_headers) do
```
```suggestion
  @spec to_raw_headers([header()], boolean()) :: [raw_header()]
  def to_raw_headers(headers, downcase_headers?) do
```

DunyaKokoschka commented:
I've added some commits which address the comments. The name of the option is currently `case_sensitive_headers`. And I added the docs for the option to the `Mint.HTTP1.connect` method. I'm not sure if this is the correct place. I wasn't so sure about this comment: [case_sensitive_headers](https://github.com/elixir-mint/mint/pull/399#discussion_r1175788370)

Also, locally the twitter integration test is failing for me with a 302 instead of a 200.

ericmj commented:
@whatyouhide I rebased and did some clean up. Do you think we are good to merge this?

wojtekmach commented on lib/mint/core/headers.ex:
```diff
@@ -0,0 +1,134 @@
+defmodule Mint.Core.Headers do
```
```suggestion
defmodule Mint.Core.Headers do
  @moduledoc false

```

wojtekmach commented on lib/mint/http1.ex:
```diff
@@ -117,14 +117,21 @@ defmodule Mint.HTTP1 do
+
+    * `:case_sensitive_headers` - (boolean) if set to true the case of the supplied
+       headers in requests will be preserved. The default is to lowercase the headers
+       because http/1.1 header names are not case-insensitive.
```
```suggestion
       because HTTP/1.1 header names are case-sensitive.
```

ericmj replied:
Oh wait. This text is even wrong, we can lower because they are in fact case-insensitive.

wojtekmach commented on lib/mint/core/headers.ex:
```diff
@@ -0,0 +1,134 @@
+          {original_name :: String.t(), canonical_name :: String.t(), value :: String.t()}
+  @type raw_header() :: {original_name :: String.t(), value :: String.t()}
+
+  @unallowed_trailer_headers MapSet.new([
```
I believe I mentioned it on some other occasion: "trailer headers" feels like a misnomer. Quickly looking at RFC 7231 the naming didn't seem to be super explicit, but in the newer one that surpasses it, RFC 9110, it is: http has _fields_: header fields (colloquially "headers") and trailer fields (colloquially "trailers")

I only mention this as an offer to send a patch to use "trailers" or "trailer fields" naming internally. It's probably not worth it to go through deprecations if the trailer_headers name was used in public APIs so I wouldn't touch these.

Feel free to ignore, it's not a big deal in the grand scheme of things. :)

whatyouhide commented on lib/mint/core/headers.ex:
```diff
@@ -0,0 +1,134 @@
+    List.keymember?(headers, name, 1)
+  end
+
+  @spec put_new_header([canonical_header()], String.t(), String.t(), String.t() | nil) ::
+          [canonical_header()]
+  def put_new_header(headers, _name, _canonical_name, nil) do
```
Maybe this is enough given the new module name?

```suggestion
  @spec put_new([canonical_header()], String.t(), String.t(), String.t() | nil) ::
          [canonical_header()]
  def put_new(headers, _name, _canonical_name, nil) do
```

whatyouhide replied:
Same for all other functions, we can prolly remove the `header` part of the name?

ericmj replied:
Definitely, good call.

whatyouhide commented on lib/mint/core/headers.ex:
```diff
@@ -0,0 +1,134 @@
+
+  @spec from_raw_headers([raw_header()]) :: [canonical_header()]
+  def from_raw_headers(headers) do
+    Enum.map(headers, fn {name, value} -> {name, String.downcase(name, :ascii), value} end)
```
Tiny, but since this is so low level, wanna go with `lists:map/2` for Performance™?

whatyouhide replied:
Also, to make sure we don't fuck this up in the future:

```suggestion
    Enum.map(headers, fn {name, value} -> {name, lower_raw(name), value} end)
```

?

whatyouhide reviewed (approved):
Beautiful Eric, couple of comments but we can finally ship this, exciting!!
